gcloud/monitoring/query.py

```diff
@@ -0,0 +1,661 @@
+    :param end_time: The end time (inclusive) of the time interval
+        for which results should be returned, as a datetime object.
+        The default is the start of the current minute.
+
+        The start time (exclusive) is determined by combining the
+        values of  ``days``, ``hours``, and ``minutes``, and
+        subtracting the resulting duration from the end time.
+
+        It is also allowed to omit the end time and duration here,
+        in which case
+        :meth:`~gcloud.monitoring.query.Query.select_interval`
+        must be called before the query is executed.
+
+    :type days: integer
+    :param days: The number of days in the time interval.
+
+    :type hours: integer
+    :param hours: The number of hours in the time interval.
+
+    :type minutes: integer
+    :param minutes: The number of minutes in the time interval.
+
+    :raises: :exc:`ValueError` if ``end_time`` is specified but
+        ``days``, ``hours``, and ``minutes`` are all zero.
+        If you really want to specify a point in time, use
+        :meth:`~gcloud.monitoring.query.Query.select_interval`.
+
+    .. _supported metrics: https://cloud.google.com/monitoring/api/metrics
+    """
+
+    DEFAULT_METRIC_TYPE = 'compute.googleapis.com/instance/cpu/utilization'
+
+    def __init__(self, client,
+                 metric_type=DEFAULT_METRIC_TYPE,
+                 end_time=None, days=0, hours=0, minutes=0):
+        start_time = None
+        if days or hours or minutes:
+            if end_time is None:
+                end_time = _NOW().replace(second=0, microsecond=0)
+            start_time = end_time - datetime.timedelta(days=days,
+                                                       hours=hours,
+                                                       minutes=minutes)
+        elif end_time is not None:
+            raise ValueError('Non-zero duration required for time interval.')
+
+        self._client = client
+        self._end_time = end_time
+        self._start_time = start_time
+        self._filter = _Filter(metric_type)
+
+        self._per_series_aligner = None
+        self._alignment_period_seconds = None
+        self._cross_series_reducer = None
+        self._group_by_fields = ()
+
+    def __iter__(self):
+        return self.iter()
+
+    @property
+    def metric_type(self):
+        """The metric type name."""
+        return self._filter.metric_type
+
+    @property
+    def filter(self):
+        """The filter string.
+
+        This is constructed from the metric type, the resource type, and
+        selectors for the group ID, monitored projects, resource labels,
+        and metric labels.
+        """
+        return str(self._filter)
+
+    def select_interval(self, end_time, start_time=None):
+        """Copy the query and set the query time interval.
+
+        Example::
+
+            import datetime
+
+            now = datetime.datetime.utcnow()
+            query = query.select_interval(
```

Comment:
I suppose it's pretty clear, given that there is no public interface for mutating them. The docstrings for the methods are very explicit about returning a copy.